File: server/src/shared/session.ts
import { createHmac, createHash, timingSafeEqual } from 'node:crypto';
import { loadServerConfig } from './config';

function getSigningKey(): Buffer {
  const secret = loadServerConfig().PROVIDER_SECRET_KEY;
  return createHash('sha256').update(secret).digest();
}

function computeHmac(userId: string, expiresAt: number): Buffer {
  return createHmac('sha256', getSigningKey())
    .update(`${userId}.${expiresAt}`)
    .digest();
}

export function signSessionToken(userId: string, ttlSeconds: number = 86400): string {
  const expiresAt = Math.floor(Date.now() / 1000) + ttlSeconds;
  const signature = computeHmac(userId, expiresAt);
  return `${userId}.${expiresAt}.${signature.toString('base64')}`;
}

export function verifySessionToken(token: string): string | null {
  const parts = token.split('.');
  if (parts.length !== 3) {
    return null;
  }

  const [userId, expiresAtStr, signatureB64] = parts;

  const expiresAt = Number(expiresAtStr);
  if (!Number.isFinite(expiresAt) || expiresAt <= 0) {
    return null;
  }

  if (Math.floor(Date.now() / 1000) > expiresAt) {
    return null;
  }

  try {
    const expected = computeHmac(userId, expiresAt);
    const actual = Buffer.from(signatureB64, 'base64');

    if (actual.length !== expected.length) {
      return null;
    }

    if (!timingSafeEqual(actual, expected)) {
      return null;
    }

    return userId;
  } catch {
    return null;
  }
}

export function isSignedToken(value: string): boolean {
  return value.split('.').length === 3;
}

File: python-backend/app/core/session.py
from __future__ import annotations

import base64
import hashlib
import hmac
import time


def _get_signing_key(secret: str) -> bytes:
    return hashlib.sha256(secret.encode("utf-8")).digest()


def _compute_hmac(user_id: str, expires_at: int, secret: str) -> bytes:
    return hmac.new(
        _get_signing_key(secret),
        f"{user_id}.{expires_at}".encode(),
        hashlib.sha256,
    ).digest()


def sign_session_token(user_id: str, secret: str, ttl_seconds: int = 86400) -> str:
    expires_at = int(time.time()) + ttl_seconds
    signature = _compute_hmac(user_id, expires_at, secret)
    return f"{user_id}.{expires_at}.{base64.b64encode(signature).decode('utf-8')}"


def verify_session_token(token: str, secret: str) -> str | None:
    parts = token.split(".")
    if len(parts) != 3:
        return None

    user_id, expires_at_raw, signature_b64 = parts
    try:
        expires_at = int(expires_at_raw)
    except ValueError:
        return None

    if expires_at <= 0 or int(time.time()) > expires_at:
        return None

    try:
        expected = _compute_hmac(user_id, expires_at, secret)
        actual = base64.b64decode(signature_b64.encode("utf-8"))
    except Exception:
        return None

    if len(actual) != len(expected):
        return None

    if not hmac.compare_digest(actual, expected):
        return None

    return user_id


def is_signed_token(value: str) -> bool:
    return value.count(".") == 2
